[ZEPPELIN-1720] Adding tests to verify behaviour of dynamic forms

guicaro opened the pull request:
### What is this PR for?
Adding Selenium tests to ensure proper behaviour of dynamic forms. 

### What type of PR is it?
Test

### Todos
N/A

### What is the Jira issue?
[ZEPPELIN-1720](https://issues.apache.org/jira/browse/ZEPPELIN-1720)

### How should this be tested?

1. Once should first get Firefox v. 41 as it is the latest version that works with the current version of Selenium in Apache Zeppelin.
2. You can then run the tests with following command:

`TEST_SELENIUM="true" mvn package -DfailIfNoTests=false -pl 'zeppelin-interpreter,zeppelin-zengine,zeppelin-server' -Dtest=ParagraphActionsIT
`
### Screenshots (if appropriate)
N/A

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

AhyoungRyu commented:
Cool!

1ambda commented:
Hi @randerzander. Thanks for contribution.

- Zeppelin build system relies on travis. So please setup your travis account and travis project for Zeppelin. 
- https://zeppelin.apache.org/contribution/contributions.html#continuous-integration
- The repository name must be `zeppelin` (not `incubator-zeppelin`, ...)

guicaro commented:
@1ambda Thanks, I actually did setup Travis while I was developing, I did not realize that every commit triggered a build so I deactivated my Zeppelin fork on Travis after I noticed this (so only have 2 builds on Travis). I activated it before I got this PR, so now I am in a state where I don't have a build for my latest commit and this PR does not show up on Travis. Should I trigger a manual build of last commit via the Travis API? Is there any other way to get back on track?

1ambda commented:
@guicaro I see. There are many ways to trigger travis again. but what you actually need is pushing new commits so that travis get noticed. 

- Amend your last commit with `git commit --amend` and `push --force`

guicaro commented:
Thanks again @1ambda it all worked like a charm and now all green :)

AhyoungRyu commented on zeppelin-server/src/test/java/org/apache/zeppelin/integration/ParagraphActionsIT.java:
```diff
@@ -483,4 +483,149 @@ public void testEditOnDoubleClick() throws Exception {
+      driver.findElement(By.xpath(getParagraphXPath(1) + "//input")).clear();
+      driver.findElement(By.xpath(getParagraphXPath(1) + "//input")).sendKeys("Zeppelin");
+
+      collector.checkThat("After new data in text input form, output should not change",
```
A minor grammar error is in the description:
`output should not change` -> `output should not be changed`

guicaro replied:
Thanks @AhyoungRyu I went ahead and checked the other test case messages and fixed an extra white space I found.

AhyoungRyu commented on zeppelin-server/src/test/java/org/apache/zeppelin/integration/ParagraphActionsIT.java:
```diff
@@ -483,4 +483,149 @@ public void testEditOnDoubleClick() throws Exception {
+
+      WebElement firstCheckbox = driver.findElement(By.xpath(getParagraphXPath(1) + "//input[1]"));
+      firstCheckbox.click();
+      collector.checkThat("After unchecking one of the boxes, output should  not display any change",
```
A extra space is between `should` and `display` :D

AhyoungRyu commented on zeppelin-server/src/test/java/org/apache/zeppelin/integration/ParagraphActionsIT.java:
```diff
@@ -483,4 +483,149 @@ public void testEditOnDoubleClick() throws Exception {
+
+      runParagraph(1);
+      waitForParagraph(1, "FINISHED");
+      collector.checkThat("Only after paragraph run we should see the new output without the box we unchecked",
```
How about "Only after paragraph run we should see the new output ~ " -> "Only after running the paragraph, we can see the newly updated output~" ?

AhyoungRyu commented on zeppelin-server/src/test/java/org/apache/zeppelin/integration/ParagraphActionsIT.java:
```diff
@@ -483,4 +483,149 @@ public void testEditOnDoubleClick() throws Exception {
+
+      runParagraph(1);
+      waitForParagraph(1, "FINISHED");
+      collector.checkThat("Only after paragraph run we should see the new output including the selected option",
```
"Only after paragraph run we should see the new output ~" -> "Only after running the paragraph, we can see the new output ~"

AhyoungRyu commented:
@guicaro Thanks for your great contribution!
I left some comments. Could you take a look at them? Please let me know if there is something that I missed :)

AhyoungRyu commented on zeppelin-server/src/test/java/org/apache/zeppelin/integration/ParagraphActionsIT.java:
```diff
@@ -483,4 +483,149 @@ public void testEditOnDoubleClick() throws Exception {
+      deleteTestNotebook(driver);
+
+    } catch (Exception e) {
+      handleException("Exception in ParagraphActionsIT while testSingleDynamicFormTextInput  ", e);
```
`testSingleDynamicFormTextInput` -> `testMultipleDynamicFormsSameType `

AhyoungRyu commented on zeppelin-server/src/test/java/org/apache/zeppelin/integration/ParagraphActionsIT.java:
```diff
@@ -483,4 +483,149 @@ public void testEditOnDoubleClick() throws Exception {
+      deleteTestNotebook(driver);
+
+    } catch (Exception e) {
+      handleException("Exception in ParagraphActionsIT while testSingleDynamicFormTextInput  ", e);
```
`testSingleDynamicFormTextInput ` -> `testSingleDynamicFormCheckboxForm `

AhyoungRyu commented on zeppelin-server/src/test/java/org/apache/zeppelin/integration/ParagraphActionsIT.java:
```diff
@@ -483,4 +483,149 @@ public void testEditOnDoubleClick() throws Exception {
+      deleteTestNotebook(driver);
+
+    } catch (Exception e) {
+      handleException("Exception in ParagraphActionsIT while testSingleDynamicFormTextInput  ", e);
```
`testSingleDynamicFormTextInput ` -> `testSingleDynamicFormSelectForm `

AhyoungRyu commented on zeppelin-server/src/test/java/org/apache/zeppelin/integration/ParagraphActionsIT.java:
```diff
@@ -483,4 +483,149 @@ public void testEditOnDoubleClick() throws Exception {
+      dropDownMenu.selectByVisibleText("Alice");
+      collector.checkThat("After selection in drop down menu, output should not display any of the options",
+              driver.findElement(By.xpath(getParagraphXPath(1) + "//div[contains(@class, 'text plainTextContent')]")).getText(),
+              CoreMatchers.equalTo("Howdy "));
```
Can you check this test case again? If we select an option in dropdown menu, the paragraph will be running and the result will be updated accordingly. So the expected result in here is `Howdy 1` instead of  just`Howdy`.  

The below screenshot img is the result of this test case when I run this in my local. 
<img width="850" alt="screen shot 2017-03-17 at 5 09 07 pm" src="https://cloud.githubusercontent.com/assets/10060731/24035322/7cd39de0-0b38-11e7-968b-58bf0da1afee.png">

guicaro replied:
Is this the result of #2060 ? Are paragraphs now run automatically by default when they contain a select form (unless one unchecks the auto-run box?

AhyoungRyu replied:
Yeah right.. You can check the updated behaviour and your test result after you rebase from master.

guicaro replied:
Refactored testSingleDynamicFormSelectForm, please tell me what you think.

1. After adding a select form I make a selection and expect the paragraph to run automatically after selection is made. 
2. I uncheck the 'Run on selection change' check box and make another selection but this time not expecting the output to change.

Also, I'm not very proud of lines 578 and 579, not sure if there is another way to reference the same select form in Selenium even after the DOM changes (due to the settings popping up).

AhyoungRyu commented:
Seems some review comments are mixed up in github side :( 
I just removed some of them and left comment again. Please let me know if you're confused.

guicaro commented:
@AhyoungRyu I'm currently looking at why testMultipleDynamicFormsSameType is failing. It now fails after I rebased from master. It's the test case that includes two select forms in one single paragraph. Basically the select foms are not showing up in GUI. I tested this with new version of Chrome (Version 56.0.2924.87 (64-bit)) to see if this might be an issue with the old Firefox version we use for Selenium, but also happens in Chrome. Here is the pic:

![seleniumchrome](https://cloud.githubusercontent.com/assets/1750596/24071511/badd04a8-0be4-11e7-88f5-f10a715458ca.JPG)

You can reproduce by entering the following string into a paragraph and then running (this used to work before I rebased).

`%spark val options = Seq(("han","Han"), ("leia","Leia"), ("luke","Luke")); println("Greetings "+z.checkbox("skywalkers", options).mkString(" and "))`

guicaro commented:
Quick question, what is the behavior when we have more than 2 select forms? Shouldn't we also have a **"Run on selection"** check box? I believe this was one of the pain points for some (see [ZEPPELIN-1013](https://issues.apache.org/jira/browse/ZEPPELIN-1013)). Here is a screen shot showing a paragraph with 2 forms created programmatically but not showing the drop down menus nor the 'Run on selection" checkbox:

![selenium2](https://cloud.githubusercontent.com/assets/1750596/24071561/28b26ca6-0be6-11e7-9922-6d0014b8f0d1.JPG)

Maybe I'm missing something? I rebased yesterday from master and have changes up to March 17 on [my fork ](https://github.com/guicaro/zeppelin/commits/master)

AhyoungRyu commented:
@guicaro Thanks for letting me know. Let me check again.

AhyoungRyu commented:
@guicaro I feel so sorry to you since it was my bad. Just created hotfix patch #2154 to fix below two issues that you reported.
 - Select form is not shown up when multiple dynamic forms (e.g. 2 select forms / 1 checkbox + 1 selectform) are in one paragraph
 - Can't see `Run on selection change` menu under the paragraph control menu

It'll be appreciated if you can help me out to review #2154!

BTW I tested all your test cases on my patch, most of them passed except `testMultipleDynamicFormsSameType`. It's because of [this line](https://github.com/apache/zeppelin/pull/2141/files#diff-48409cf8ca5ddb1aea1a62ada3cd091dR649). The expected matcher should be `"Howdy 1\nHowdy "` instead of `"Howdy \nHowdy "` since the paragraph runs right after the selection has been changed. (as I said before :D)

And after you rebased, some irrelevant commits are mixed and pushed together in here. Could you handle them? Feel free to ping me if you're in trouble while you resolve it.

guicaro commented:
@AhyoungRyu  no worries, will be glad to review your work, fix `testMultipleDynamicFormsSameType` and fox my commits, will definetly ping you about this last issue :) Thanks

AhyoungRyu commented:
@guicaro There are some unrelated commits in here after you rebased. Can you try to fix it?

guicaro commented:
@AhyoungRyu Fixed the issue with my commits, big shoutout to @DrIgor for helping resolve my issue! I also updated the tests (ex. `testSingleDynamicFormSelectForm`, `testSingleDynamicFormCheckboxForm`, `testMultipleDynamicFormsSameType`) to work with the new paragraph behaviour. All 14 tests pass now with the latest code in master.

guicaro commented:
Selenium tests are red in Travis but due to:

`SparkParagraphIT.testSqlSpark:185 » NoSuchElement Unable to locate element: {"... `

https://travis-ci.org/guicaro/zeppelin/jobs/212973588

Leemoonsoo commented:
@guicaro Thanks for the contribution!
CI problem has been fixed on master branch. Could you try rebase or merge master branch and see if CI becomes green?

guicaro commented:
@Leemoonsoo rebased from master and all green now, thanks for the heads up!

AhyoungRyu commented:
@guicaro Thanks for such great work! LGTM

AhyoungRyu commented:
Merge into master and branch-0.7 if there are no more comments on this.
